File: backend/encryption/symmetric/aes_manual.py
import base64
import os
from .base import SymmetricEncryption

class AESManual(SymmetricEncryption):
    def __init__(self):
        self.block_size = 16

    def _generate_dynamic_sbox(self, key: bytes):
        sbox = list(range(256))
        # JS tarafıyla birebir aynı toplam sonucunu almak için % 256 eklenmiştir
        seed = sum(list(key)) % 256 
        for i in range(255, 0, -1):
            j = (seed + i) % (i + 1)
            sbox[i], sbox[j] = sbox[j], sbox[i]
            seed = (seed + sbox[i]) % 256
        return sbox

    def encrypt(self, plaintext, key: bytes) -> str:
        sbox = self._generate_dynamic_sbox(key)
        iv = os.urandom(16)
        if isinstance(plaintext, str):
            data = self._pad(plaintext.encode('utf-8'))
        else:
            data = self._pad(plaintext)
            
        res = bytearray()
        for i in range(len(data)):
            # S-Box -> Key XOR -> IV XOR sırası
            val = sbox[data[i]] ^ key[i % len(key)] ^ iv[i % 16]
            res.append(val)
        # IV ve şifreli veriyi birleştirip base64 yapıyoruz
        return base64.b64encode(iv + res).decode('utf-8')

    def decrypt(self, ciphertext: str, key: bytes):
        sbox = self._generate_dynamic_sbox(key)
        inv_sbox = [0] * 256
        for i, v in enumerate(sbox): inv_sbox[v] = i
        
        combined = base64.b64decode(ciphertext.encode('utf-8'))
        iv, data = combined[:16], combined[16:]
        res = bytearray()
        for i in range(len(data)):
            # XOR işlemlerini tersine alıyoruz
            val = data[i] ^ key[i % len(key)] ^ iv[i % 16]
            res.append(inv_sbox[val])
        
        decrypted_bytes = self._unpad(bytes(res))
        try:
            return decrypted_bytes.decode('utf-8')
        except UnicodeDecodeError:
            return decrypted_bytes

    def _pad(self, data: bytes) -> bytes:
        p = self.block_size - (len(data) % self.block_size)
        return data + bytes([p] * p)

    def _unpad(self, data: bytes) -> bytes:
        if not data: return b""
        p = data[-1]
        if 0 < p <= self.block_size: return data[:-p]
        return data

    def generate_key(self, key_size: int = 128) -> bytes:
        return os.urandom(key_size // 8)

File: frontend/static/js/encryption/symmetric/aes_manual.js
class AESManual {
    generateDynamicSBox(key) {
        let sbox = Array.from({length: 256}, (_, i) => i);
        let seed = Array.from(key).reduce((a, b) => a + b, 0) % 256;
        for (let i = 255; i > 0; i--) {
            let j = (seed + i) % (i + 1);
            [sbox[i], sbox[j]] = [sbox[j], sbox[i]];
            seed = (seed + sbox[i]) % 256;
        }
        return sbox;
    }

    async encrypt(plaintext, key) {
        const sbox = this.generateDynamicSBox(key);
        const iv = window.crypto.getRandomValues(new Uint8Array(16));
        const data = this.pad(new TextEncoder().encode(plaintext));
        const res = new Uint8Array(data.length);
        
        for (let i = 0; i < data.length; i++) {
            let val = sbox[data[i]];
            res[i] = val ^ key[i % key.length] ^ iv[i % 16];
        }

        const combined = new Uint8Array(iv.length + res.length);
        combined.set(iv);
        combined.set(res, iv.length);
        return this.arrayBufferToBase64(combined);
    }

    async decrypt(ciphertext, key) {
        const sbox = this.generateDynamicSBox(key);
        const invSbox = new Array(256);
        for(let i=0; i<256; i++) invSbox[sbox[i]] = i;

        const combined = this.base64ToArrayBuffer(ciphertext);
        const iv = combined.slice(0, 16);
        const data = combined.slice(16);
        const res = new Uint8Array(data.length);

        for (let i = 0; i < data.length; i++) {
            let val = data[i] ^ key[i % key.length] ^ iv[i % 16];
            res[i] = invSbox[val % 256];
        }
        return new TextDecoder().decode(this.unpad(res));
    }

    pad(d) {
        const p = 16 - (d.length % 16);
        const padded = new Uint8Array(d.length + p);
        padded.set(d);
        for (let i = d.length; i < padded.length; i++) padded[i] = p;
        return padded;
    }

    unpad(d) {
        const p = d[d.length - 1];
        return d.slice(0, d.length - p);
    }

    arrayBufferToBase64(uint8) {
        let binary = '';
        uint8.forEach(b => binary += String.fromCharCode(b));
        return btoa(binary);
    }

    base64ToArrayBuffer(base64) {
        const binary = atob(base64);
        return Uint8Array.from(binary, c => c.charCodeAt(0));
    }

    async generateKey() { return window.crypto.getRandomValues(new Uint8Array(16)); }
}